### ISSUE TITLE
NLog sometimes creates thousands of empty archive files when using DateAndSequence archiving

### ISSUE THREAD

**Issue by f-i-x-7** [NLog/NLog#4860]
### Bug template

**NLog version**: 4.7.12; if it is important - we do not use NLog directly but rather use NLog.Extensions.Logging 1.7.4 and Microsoft.Extensions.Logging 3.1.22

**Platform**: .NET Core 3.1

**Current NLog config** (xml or C#, if relevant)

```xml
<?xml version="1.0" encoding="utf-8" ?>
<nlog xmlns="http://www.nlog-project.org/schemas/NLog.xsd"
      xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
      xsi:schemaLocation="http://www.nlog-project.org/schemas/NLog.xsd NLog.xsd"
      autoReload="true">
  <variable name="baseLayout" value="${date:format=yyyy\-MM\-dd HH\:mm\:ss\.fff}|${level}|${threadid}|${logger}${when:when='${ndlc}' != '':inner=|${ndlc}}|${message}" />
  <variable name="exceptionLayoutSuffix" value="&#13;&#10;${exception:format=ToString,Data}" />
  <variable name="defaultLayout" value="${var:baseLayout}${onexception:inner=${var:exceptionLayoutSuffix}}" />
  <variable name="logDirectory" value="${configsetting:name=LoggingParams.Directory}"/> <!-- value taken from appsettings.json -->
  
  <targets>
    <target name="file" xsi:type="AsyncWrapper" overflowAction="Grow">
      <target xsi:type="File" fileName="${logDirectory}/MyApp.log"
        archiveEvery="Day"
        archiveFileName="${logDirectory}/Archives/MyApp_{#}.zip"
        archiveDateFormat="yyyy-MM-dd"
        archiveNumbering="DateAndSequence"
        archiveAboveSize="314572800"
        maxArchiveFiles="3000"
        enableArchiveFileCompression="true"
        layout="${var:defaultLayout}" />
    </target>
    <target name="console" xsi:type="AsyncWrapper" overflowAction="Grow">
      <target xsi:type="ColoredConsole" layout="${var:defaultLayout}" />
    </target>
  </targets>
  <rules>
    <logger name="*" minlevel="Debug" writeTo="file,console" />
  </rules>
</nlog>
```

- What is the current result?

Sometimes (far-far from regular) thousands of empty archive files with big sequence numbers are created. But sequence numbers do not start from 0. ![issue](https://user-images.githubusercontent.com/64615981/161022554-e2149ad7-7cbb-4b2e-9217-96ec0132b2e6.jpg)

Also it seems that archives with real data (I mean not empty archives with actual compressed log files) are not created or lost (but not 100% sure about this).

Such behavior was observed about 10 times during approximately 5 months while app was running at 3 servers.

- What is the expected result?

No thousands of empty archive files

- Did you checked the [Internal log](https://github.com/NLog/NLog/wiki/Internal-Logging)?

No

- Please post full exception details (message, stacktrace, inner exceptions)

No exception was observed

- Are there any workarounds? yes/no

Don't know, I only can guess about reasons of this (see comments below)

- Is there a version in which it did work?

Don't know (may be issue is related with specific configuration details - see my comments below - and such configuration was applied  by our team only at 4.7.12 NLog version)

- Can you help us by writing an unit test?

I can try to write a repro app that actively writes logs and uses such configuration, but I have doubts that the bug can be reproduced within reasonable amount of time (I judge by the bug frequency observed in real-world usage; but may be lower value of archive size threshold can help)

### Additional thoughts

I've read the docs and I suspect the bug is related with Dynamic FileName Archive Logic and Static FileName Archive Logic mixing. As it is said [here](https://github.com/NLog/NLog/wiki/File-target#dynamic-vs-static-archive-logic):

`Do not mix "Dynamic FileName Archive Logic" together with "Static FileName Archive Logic" as one will get unexpected archive behavior.`

It seems there are both Dynamic FileName Archive Logic (because of `${logDirectory}` in `fileName`) and Static FileName Archive Logic (because of  `archiveEvery="Day"` and initialized value of `archiveFileName`). Am I right?

On the other hand, the article cited above links to [another article with examples](https://github.com/NLog/NLog/wiki/FileTarget-Archive-Examples), and [the very first example](https://github.com/NLog/NLog/wiki/FileTarget-Archive-Examples#size-based-file-archival) seems to be mixing Dynamic and Static FileName archive logic as well:
- `${basedir}` is used in `fileName` => Dynamic FileName Archive Logic is used;
- `archiveFileName` has initialized value => Static FileName Archive Logic is used.

I wanted to remove `${logDirectory}` from my app config and test app for a while, but, given the mentioned example, now I'm a bit confused whether I am right regarding the root cause.

**Comment by f-i-x-7** [comment]
> Also it seems that archives with real data (I mean not empty archives with actual compressed log files) are not created or lost (but not 100% sure about this).

Ah, that's because of `maxArchiveFiles="3000"`

**Comment by snakefoot** [comment]
Closing due to inactivity. If new input arrive, then don't hesitate to re-open.

**Comment by f-i-x-7** [comment]
I would like to reopen issue as there are now more details.

After enabling NLog internal logs on our test environment, I hadn't much free time to review accumulated logs. But recently I finally checked them, and I noticed that on 2022-04-14 the issue reproduced on 2 servers (not exactly at the same time, but pretty close).

When plenty of empty archives was created then following exception was written into logs frequently:

```
2022-04-14 11:13:17.5547 Error FileTarget(Name=): Failed write to file 'C:\PAPA\XXX.Algo.Manager.Service\Current\../../Logs/XXX.Algo.Manager.Service/XXX.Algo.Manager.Service.log'. Exception: System.UnauthorizedAccessException: Access to the path 'C:\PAPA\Logs\XXX.Algo.Manager.Service\XXX.Algo.Manager.Service.log' is denied.
   at System.IO.FileStream.ValidateFileHandle(SafeFileHandle fileHandle)
   at System.IO.FileStream.CreateFileOpenHandle(FileMode mode, FileShare share, FileOptions options)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options)
   at NLog.Internal.FileAppenders.BaseFileAppender.TryCreateFileStream(Boolean allowFileSharedWriting, Int32 overrideBufferSize)
   at NLog.Internal.FileAppenders.BaseFileAppender.CreateFileStream(Boolean allowFileSharedWriting, Int32 overrideBufferSize)
   at NLog.Internal.FileAppenders.RetryingMultiProcessFileAppender.Write(Byte[] bytes, Int32 offset, Int32 count)
   at NLog.Targets.FileTarget.WriteToFile(String fileName, ArraySegment`1 bytes, Boolean initializedNewFile)
```

NLog internal logs for both servers are attached.
[NLog_issue.zip](https://github.com/NLog/NLog/files/8800570/NLog_issue.zip)
Please note that I removed logs that were written before or after this date.

The issue on 1st server lasted from 11:13 till 11:18, on 2nd - from 10:53 till 10:55.

P.S. According to these logs, there were other errors at 10:59 on 1st server, but I don't think that this is related to the issue being opened.

**Comment by snakefoot** [comment]
Looks the application is fighting other applications for writing to the file:
```
2022-04-14 10:55:55.4412 Debug FileTarget(Name=): Start archiving 'C:\PAPA\XXX.Algo.Manager.Service\Current\../../Logs/XXX.Algo.Manager.Service/XXX.Algo.Manager.Service.log' because FileSize=314284156 + 298723 is larger than ArchiveAboveSize=314572800
2022-04-14 10:55:55.4412 Debug FileAppender Closing Invalidate - C:\PAPA\XXX.Algo.Manager.Service\Current\../../Logs/XXX.Algo.Manager.Service/XXX.Algo.Manager.Service.log
2022-04-14 10:55:55.4601 Debug FileTarget(Name=): Start archiving 'C:\PAPA\XXX.Algo.Manager.Service\Current\../../Logs/XXX.Algo.Manager.Service/XXX.Algo.Manager.Service.log' because FileSize=314284156 + 298723 is larger than ArchiveAboveSize=314572800
2022-04-14 10:55:55.4601 Debug FileTarget(Name=): Do archive file: 'C:\PAPA\XXX.Algo.Manager.Service\Current\../../Logs/XXX.Algo.Manager.Service/XXX.Algo.Manager.Service.log'
2022-04-14 10:55:55.4724 Info FileTarget(Name=): Archiving C:\PAPA\Logs\XXX.Algo.Manager.Service\XXX.Algo.Manager.Service.log to compressed C:\PAPA\Logs\Archives\XXX.Algo.Manager.Service\XXX.Algo.Manager.Service_2022-04-14.0.zip
2022-04-14 10:55:57.4410 Info FileTarget(Name=): Deleting old archive file: 'C:\PAPA\Logs\XXX.Algo.Manager.Service\XXX.Algo.Manager.Service.log'.
2022-04-14 10:55:57.5349 Debug Creating file appender: C:\PAPA\XXX.Algo.Manager.Service\Current\../../Logs/XXX.Algo.Manager.Service/XXX.Algo.Manager.Service.log
2022-04-14 10:59:22.6598 Warn Attempt #0 to open C:\PAPA\XXX.Algo.Manager.Service\Current\../../Logs/XXX.Algo.Manager.Service/XXX.Algo.Manager.Service.log failed. Sleeping for 0ms
2022-04-14 10:59:22.6598 Warn Attempt #1 to open C:\PAPA\XXX.Algo.Manager.Service\Current\../../Logs/XXX.Algo.Manager.Service/XXX.Algo.Manager.Service.log failed. Sleeping for 0ms
2022-04-14 10:59:22.6670 Warn Attempt #2 to open C:\PAPA\XXX.Algo.Manager.Service\Current\../../Logs/XXX.Algo.Manager.Service/XXX.Algo.Manager.Service.log failed. Sleeping for 0ms
2022-04-14 10:59:22.6670 Warn Attempt #3 to open C:\PAPA\XXX.Algo.Manager.Service\Current\../../Logs/XXX.Algo.Manager.Service/XXX.Algo.Manager.Service.log failed. Sleeping for 1ms
2022-04-14 10:59:22.6825 Warn Attempt #4 to open C:\PAPA\XXX.Algo.Manager.Service\Current\../../Logs/XXX.Algo.Manager.Service/XXX.Algo.Manager.Service.log failed. Sleeping for 8ms
2022-04-14 10:59:22.6825 Warn Attempt #5 to open C:\PAPA\XXX.Algo.Manager.Service\Current\../../Logs/XXX.Algo.Manager.Service/XXX.Algo.Manager.Service.log failed. Sleeping for 0ms
2022-04-14 10:59:22.7048 Warn Attempt #6 to open C:\PAPA\XXX.Algo.Manager.Service\Current\../../Logs/XXX.Algo.Manager.Service/XXX.Algo.Manager.Service.log failed. Sleeping for 39ms
2022-04-14 10:59:22.7498 Warn Attempt #7 to open C:\PAPA\XXX.Algo.Manager.Service\Current\../../Logs/XXX.Algo.Manager.Service/XXX.Algo.Manager.Service.log failed. Sleeping for 106ms
2022-04-14 10:59:22.8630 Warn Attempt #8 to open C:\PAPA\XXX.Algo.Manager.Service\Current\../../Logs/XXX.Algo.Manager.Service/XXX.Algo.Manager.Service.log failed. Sleeping for 48ms
2022-04-14 10:59:22.9180 Error FileTarget(Name=): Failed write to file 'C:\PAPA\XXX.Algo.Manager.Service\Current\../../Logs/XXX.Algo.Manager.Service/XXX.Algo.Manager.Service.log'. Exception: System.IO.IOException: The process cannot access the file 'C:\PAPA\Logs\XXX.Algo.Manager.Service\XXX.Algo.Manager.Service.log' because it is being used by another process.
   at System.IO.FileStream.ValidateFileHandle(SafeFileHandle fileHandle)
   at System.IO.FileStream.CreateFileOpenHandle(FileMode mode, FileShare share, FileOptions options)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options)
   at NLog.Internal.FileAppenders.BaseFileAppender.TryCreateFileStream(Boolean allowFileSharedWriting, Int32 overrideBufferSize)
   at NLog.Internal.FileAppenders.BaseFileAppender.CreateFileStream(Boolean allowFileSharedWriting, Int32 overrideBufferSize)
   at NLog.Internal.FileAppenders.RetryingMultiProcessFileAppender.Write(Byte[] bytes, Int32 offset, Int32 count)
   at NLog.Targets.FileTarget.WriteToFile(String fileName, ArraySegment`1 bytes, Boolean initializedNewFile)
2022-04-14 10:59:22.9180 Debug FileAppender Closing Invalidate - C:\PAPA\XXX.Algo.Manager.Service\Current\../../Logs/XXX.Algo.Manager.Service/XXX.Algo.Manager.Service.log
```

Then it complains about access denied (Just after this "Timeout"):
```
2022-04-14 11:13:17.5385 Warn FileTarget(Name=): Timeout while deleting old archive file: 'C:\PAPA\Logs\XXX.Algo.Manager.Service\XXX.Algo.Manager.Service.log'.
2022-04-14 11:13:17.5385 Debug FileTarget(Name=): Preparing for new file 'C:\PAPA\XXX.Algo.Manager.Service\Current\../../Logs/XXX.Algo.Manager.Service/XXX.Algo.Manager.Service.log'
2022-04-14 11:13:17.5385 Debug Creating file appender: C:\PAPA\XXX.Algo.Manager.Service\Current\../../Logs/XXX.Algo.Manager.Service/XXX.Algo.Manager.Service.log
2022-04-14 11:13:17.5547 Error FileTarget(Name=): Failed write to file 'C:\PAPA\XXX.Algo.Manager.Service\Current\../../Logs/XXX.Algo.Manager.Service/XXX.Algo.Manager.Service.log'. Exception: System.UnauthorizedAccessException: Access to the path 'C:\PAPA\Logs\XXX.Algo.Manager.Service\XXX.Algo.Manager.Service.log' is denied.
...

And the access denied continues for a long time, until the file disappears and NLog recovers:
```
2022-04-14 11:18:04.1364 Error FileTarget(Name=): Failed write to file 'C:\PAPA\XXX.Algo.Manager.Service\Current\../../Logs/XXX.Algo.Manager.Service/XXX.Algo.Manager.Service.log'. Exception: System.UnauthorizedAccessException: Access to the path 'C:\PAPA\Logs\XXX.Algo.Manager.Service\XXX.Algo.Manager.Service.log' is denied.
2022-04-14 11:18:04.1364 Debug FileAppender Closing Invalidate - C:\PAPA\XXX.Algo.Manager.Service\Current\../../Logs/XXX.Algo.Manager.Service/XXX.Algo.Manager.Service.log
2022-04-14 11:18:04.3241 Debug FileTarget(Name=): Invalidates appender for archive file 'C:\PAPA\XXX.Algo.Manager.Service\Current\../../Logs/XXX.Algo.Manager.Service/XXX.Algo.Manager.Service.log' since it no longer exists
2022-04-14 11:18:04.3241 Debug Creating file appender: C:\PAPA\XXX.Algo.Manager.Service\Current\../../Logs/XXX.Algo.Manager.Service/XXX.Algo.Manager.Service.log
2022-04-14 11:18:04.3391 Debug FileTarget(Name=): Preparing for new file 'C:\PAPA\XXX.Algo.Manager.Service\Current\../../Logs/XXX.Algo.Manager.Service/XXX.Algo.Manager.Service.log'
```

**Comment by snakefoot** [comment]
I'm guessing some other application is holding the file open

**Comment by snakefoot** [comment]
NLog v5 changes the default value to [KeepFileOpen=true](https://nlog-project.org/2021/08/25/nlog-5-0-preview1-ready.html#filetarget-keepfileopen--true-by-default), because of the issues with background-applications (ex. malware scanners, file-viewers, etc.) taking over the file-handles and causing havoc for the host-application.

**Comment by snakefoot** [comment]
Ofcourse interested in reducing the number of ZIP-files, when there are problems. 

Did NLog FileTarget generate a lot of ZIP-files in the period with the error?

**Comment by f-i-x-7** [comment]
Thanks for quick feedback!


> I'm guessing some other application is holding the file open, and disturbs the NLog FileTarget

In theory, it could be team members who viewed the logs, but IMHO it is much more likely that it is some background scanner from corporate security department or smth like that. This means we cannot eliminate the root cause.


> NLog v5 changes the default value to [KeepFileOpen=true](https://nlog-project.org/2021/08/25/nlog-5-0-preview1-ready.html#filetarget-keepfileopen--true-by-default), because of the issues with background-applications (ex. malware scanners, file-viewers, etc.) taking over the file-handles and causing havoc for the host-application.

Thanks for pointing out to this setting. We will investigate whether it is acceptable for us to enable `keepFileOpen=true` while still being stick to NLog 4.7.12 or higher.


> Did NLog FileTarget generate a lot of ZIP-files in the latest error-period?

Yeah, there were hundreds of empty ZIP files on each server:

1st server - from XXX.Algo.Manager.Service_2022-04-14.3.zip till XXX.Algo.Manager.Service_2022-04-14.462.zip (total 460 files)
2nd server - from XXX.Algo.Manager.Service_2022-04-14.1.zip till XXX.Algo.Manager.Service_2022-04-14.570.zip (total 570 files)

And that's interesting... The issue lasted less time on 2nd server, but number of generated ZIP files is bigger.


> Ofcourse interested in reducing the number of ZIP-files, when there are problems with file-handles.

It would be great!

**Comment by snakefoot** [comment]
Believe I have identified the cause. The NLog method CompressFile first attempts to open the zip-file and then attempts to open the file to compress.

This should probably be changed so it first aquire handle to the file to be compressed, and then creates the zip-file.

Should be easy to fix, and can be included in NLog v5.0.1



Sent from my Galaxy



-------- Original message --------
From: Artem Eliseev ***@***.***>
Date: 31/05/2022 09:46 (GMT+01:00)
To: NLog/NLog ***@***.***>
Cc: Rolf Kristensen ***@***.***>, State change ***@***.***>
Subject: Re: [NLog/NLog] NLog sometimes creates thousands of empty archive files when using DateAndSequence archiving (Issue #4860)


Thanks for quick feedback!

I'm guessing some other application is holding the file open, and disturbs the NLog FileTarget

In theory, it could be team members who viewed the logs, but IMHO it is much more likely that it is some background scanner from corporate security department or smth like that. This means we cannot eliminate the root cause.

NLog v5 changes the default value to KeepFileOpen=true<https://nlog-project.org/2021/08/25/nlog-5-0-preview1-ready.html#filetarget-keepfileopen--true-by-default>, because of the issues with background-applications (ex. malware scanners, file-viewers, etc.) taking over the file-handles and causing havoc for the host-application.

Thanks for pointing out to this setting. We will investigate whether it is acceptable for us to enable keepFileOpen=true while still being stick to NLog 4.7.12 or higher.

Did NLog FileTarget generate a lot of ZIP-files in the latest error-period?

Yeah, there were hundreds of empty ZIP files on each server:

1st server - from XXX.Algo.Manager.Service_2022-04-14.3.zip till XXX.Algo.Manager.Service_2022-04-14.462.zip (total 460 files)
2nd server - from XXX.Algo.Manager.Service_2022-04-14.1.zip till XXX.Algo.Manager.Service_2022-04-14.570.zip (total 570 files)

And that's interesting... The issue lasted less time on 2nd server, but number of generated ZIP files is bigger.

Ofcourse interested in reducing the number of ZIP-files, when there are problems with file-handles.

It would be great!

**Comment by snakefoot** [comment]
@f-i-x-7 Thank you for being persistent, and reporting back with all this nice evidence.
